Filtered_kernel/include/CGAL/Lazy.h

```diff
@@ -878,30 +835,23 @@ struct Lazy_construction_nt {
 
   BOOST_PP_REPEAT_FROM_TO(1, 6, CGAL_RESULT_NT, _)
 
-#define CGAL_NT_OPERATOR(z, n, d)                                            \
-  template<BOOST_PP_ENUM_PARAMS(n, class L)>                            \
-  typename cpp11::result_of<Lazy_construction_nt(BOOST_PP_ENUM_PARAMS(n, L))>::type \
-  operator()( BOOST_PP_ENUM(n, CGAL_LARGS, _) ) const {                      \
-    BOOST_PP_REPEAT(n, CGAL_TYPEMAP_EC, L)                                     \
-    BOOST_PP_REPEAT(n, CGAL_TYPEMAP_AC, L)                                     \
-    typedef typename boost::remove_cv< typename boost::remove_reference < \
-                                        typename cpp11::result_of< EC(BOOST_PP_ENUM_PARAMS(n, E)) >::type >::type >::type ET; \
-    typedef typename boost::remove_cv< typename boost::remove_reference < \
-                                        typename cpp11::result_of< AC(BOOST_PP_ENUM_PARAMS(n, A)) >::type >::type >::type AT; \
-    CGAL_BRANCH_PROFILER(std::string(" failures/calls to   : ") + std::string(CGAL_PRETTY_FUNCTION), tmp); \
-    Protect_FPU_rounding<Protection> P;                                 \
-    try {                                                               \
-      return new Lazy_rep_##n<AT, ET, AC, EC, To_interval<ET>, BOOST_PP_ENUM_PARAMS(n, L) >(ac, ec, BOOST_PP_ENUM_PARAMS(n, l)); \
-    } catch (Uncertain_conversion_exception&) {                          \
-      CGAL_BRANCH_PROFILER_BRANCH(tmp);                                 \
-      Protect_FPU_rounding<!Protection> P2(CGAL_FE_TONEAREST);          \
-      return new Lazy_rep_0<AT,ET,To_interval<ET> >(ec( BOOST_PP_ENUM(n, CGAL_LEXACT, _) )); \
-    }                                                                   \
-  }                                                                     \
-
-  BOOST_PP_REPEAT_FROM_TO(1, 6, CGAL_NT_OPERATOR, _)
+  template<class...L>
+  auto operator()(L const&...l) const ->
+  Lazy_exact_nt<std::remove_cv_t<std::remove_reference_t<decltype(ec(CGAL::exact(l)...))>>>
+  {
+    typedef std::remove_cv_t<std::remove_reference_t<decltype(ec(CGAL::exact(l)...))>> ET;
+    typedef std::remove_cv_t<std::remove_reference_t<decltype(ac(CGAL::approx(l)...))>> AT;
```

Comment:
Why not `std::decay_t`? Is there any risk is can be different from `std::remove_cv_t<std::remove_reference_t<T>>``?

Reply:
I don't remember. I probably used this form because there were already some `typename boost::remove_cv< typename boost::remove_reference < typename boost::result_of<` (introduced by Philipp Möller a while ago). And boost's decay used to be very different from the standard one. I also think the standard takes care to use remove_cvref in cases where the difference with decay is irrelevant. In this case, I don't think it matters which one is used (but I didn't look deeply).